include/eve/detail/function/simd/x86/swap_adjacent_group.hpp

```diff
@@ -0,0 +1,183 @@
+              v = _mm256_shufflehi_epi16(v, sf4::value);
+          return  _mm256_shufflelo_epi16(v, sf4::value);
+          }
+          else
+          {
+            auto[l,h] = v.slice();
+            return that_t( swap_adjacent_group(l,fixed<G>{}), swap_adjacent_group(h,fixed<G>{}));
+          }
+        }
+        else if constexpr( size == 4 )
+        {
+          if constexpr( std::same_as<T,float>  )  return v[pattern<1,0,3,2,5,4,7,6>];
+          else  if constexpr( current_api >= avx2    )
+          {
+            return _mm256_shuffle_epi32(v, sf4::value);
+          }
+          else
+          {
+            auto sd = bit_cast(v, as_<typename that_t::template rebind<float>>{});
+            sd = swap_adjacent_group(sd, fixed<1>{});
+            return bit_cast(sd, as(v));
+          }
+        }
+        else if constexpr( size == 8 )
+        {
+                if constexpr( std::same_as<T,double> )  return v[pattern<1,0,2,3>];
+          else  if constexpr( std::same_as<T,float>  )  return v[pattern<2,3,0,1,6,7,4,5>];
+          else  if constexpr( current_api >= avx2    )
+          {
+            return _mm256_permute4x64_epi64(v, sf4::value);
+          }
+          else
+          {
+            auto sd = bit_cast(v, as_<typename that_t::template rebind<float>>{});
+            sd = swap_adjacent_group(sd, fixed<2>{});
+            return bit_cast(sd, as(v));
+          }
+        }
+        else if constexpr( size == 16 )
+        {
+                if constexpr( std::same_as<T,double> )  return v[pattern<2,3,0,1>];
+          else  if constexpr( std::same_as<T,float>  )  return v[pattern<4,5,6,7,0,1,2,3>];
+          else  if constexpr( current_api >= avx2    )
+          {
+            return _mm256_permute4x64_epi64(v, _MM_SHUFFLE(1,0,3,2));
+          }
+          else
+          {
+            auto[l,h] = v.slice();
+            return that_t(h,l);
+          }
+        }
+      }
+      else  if constexpr( std::same_as<ABI,x86_512_> )
+      {
+        // We have perfect swizzle so LET'S ROCK'N'ROLL
+        return basic_swizzle(v, swap_adjacent_group_n<G,N::value> );
+      }
+    }
+  }
+
+  template<real_scalar_value T, typename N, x86_abi ABI, std::ptrdiff_t G>
+  EVE_FORCEINLINE logical<wide<T,N,ABI>> swap_adjacent_group_ ( EVE_SUPPORTS(cpu_)
+                                                              , logical<wide<T,N,ABI>> v, fixed<G>
+                                                              ) noexcept
+  requires(G<=N::value)
+  {
+    if constexpr(G == N::value)
+    {
+      return v;
+    }
+    else  if constexpr( !ABI::is_wide_logical )
+    {
+      // Reconstruct mask, swag then turn to mask again
+      auto const m = v.mask();
+      auto const swag = swap_adjacent_group(m, fixed<G>{});
+      return to_logical(swag);
+    }
+    else
+    {
+      // Use basic implementation via bit_cast & mask()
+      return bit_cast( swap_adjacent_group(v.mask(),fixed<G>{}), as(v));
+    }
```

Comment:
retarget cpu?
Or just the whole thing put there?

Reply:
I had an error when retargeting but it's maybe a bad EVE_SUPPORTS flag somewhere. Will fix

Reply:
Fixed